packages/google_maps_flutter/lib/src/controller.dart

```diff
@@ -144,95 +110,76 @@ class GoogleMapController extends ChangeNotifier {
       'cameraUpdate': cameraUpdate._toJson(),
     });
   }
+}
 
-  /// Adds a marker to the map, configured using the specified custom [options].
-  ///
-  /// Change listeners are notified once the marker has been added on the
-  /// platform side.
-  ///
-  /// The returned [Future] completes with the added marker once listeners have
-  /// been notified.
-  Future<Marker> addMarker(MarkerOptions options) async {
-    final MarkerOptions effectiveOptions =
-        MarkerOptions.defaultOptions.copyWith(options);
-    // TODO(amirh): remove this on when the invokeMethod update makes it to stable Flutter.
-    // https://github.com/flutter/flutter/issues/26431
-    // ignore: strong_mode_implicit_dynamic_method
-    final String markerId = await _channel.invokeMethod(
-      'marker#add',
-      <String, dynamic>{
-        'options': effectiveOptions._toJson(),
-      },
-    );
-    final Marker marker = Marker(markerId, effectiveOptions);
-    _markers[markerId] = marker;
+/// Manages lifecycle of [MarkerController] for all [Marker]s.
+///
+/// Change listeners are notified upon changes to any of the markers.
+class MarkerControllers extends ChangeNotifier {
+  MarkerControllers();
+
+  final Map<MarkerId, MarkerController> _markerControllers =
+      <MarkerId, MarkerController>{};
+
+  void update(MarkerUpdates markerUpdates) {
+    markerUpdates.markerUpdates.forEach((MarkerUpdate markerUpdate) {
```

Comment:
I think the code would be simpler if we would just use the map method channel for all markers, that way we could save the management of the controllers here and e.g when we get a `marker#onTap(markerId)` all we need to do is fetch the Marker with id `markerId` from the markers set we maintain anyway and invoke the callback.